### ISSUE TITLE
`rocker/rstudio` for apple silicon linux/arm64/v8 platform

### ISSUE THREAD

**Issue by dvillaj** [rocker-org/rocker-versioned2#144]
Hi,

I am experiencing the following error when I try to run the container on a Macbook Pro M1 (MacOS 11.2.3):

WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested

![Screenshot 2021-04-16 at 05 20 29](https://user-images.githubusercontent.com/8428613/114966729-78cc1380-9e73-11eb-892e-ad8064091f56.png)

When I try to log-in in RStudio I get the following error: Error: Incorrect or invalid username/password

![Screenshot 2021-04-16 at 05 32 38](https://user-images.githubusercontent.com/8428613/114967673-2a1f7900-9e75-11eb-8012-fee17e1e82c2.png).

I have tried with the latest version of Docker Desktop (3.3.1) and disabling/enabled experimental features.

**Comment by style206** [comment]
I have the same problem :(
I hope that the update supporting M1 Mac.

**Comment by eitsupi** [comment]
See this issue for arm64 build of RStudio.
https://github.com/rstudio/rstudio/issues/8809

In addition, even if the above issue is resolved and RStudio for arm can be installed, RStudio Package Manager (RSPM) does not provide R binary packages for arm64, so Docker images like `rocker/tidyverse` will take a very long time to build...

**Comment by eitsupi** [comment]
The upstream issue https://github.com/rstudio/rstudio/issues/8809 appears to have been registered for the 2022-02 (or 2022-03) milestone!

**Comment by echuber2** [comment]
As a stopgap measure, does the amd64 version work (maybe slowly) [with Rosetta 2 emulation enabled](https://docs.docker.com/desktop/mac/apple-silicon/)? We have a version of the rocker image here that has auth turned off for limited usage; reportedly, some people are able to run it on M1 Macs: https://github.com/PrairieLearn/PrairieLearn/tree/master/workspaces/rstudio

**Comment by eitsupi** [comment]
It appears that the arm64 daily build for jammy is now available.

https://rstudio.org/download/latest/daily/server/jammy/rstudio-server-latest-arm64.deb

I believe this can be installed with the `install_rstudio.sh` script, but so far there is no jammy-based, arm64-architecture `rocker/r-ver` here, so I have not tried it. (`rocker/r-ver:devel` is already jammy but amd64 only)
When R 4.2.2 is released, `rocker/r-ver:latest` will be built on `ubuntu:jammy`, so we will be able to try this.

https://github.com/rocker-org/rocker-versioned2/blob/ce8821d2090c0a88ca2875582c4946536df67614/scripts/install_rstudio.sh#L60-L61

**Comment by eitsupi** [comment]
The arm64 version of RStudio daily builds are now available for focal (https://github.com/rstudio/rstudio/issues/8809#issuecomment-1224856044).
The following Dockerfile should work on arm64 platforms now!

```dockerfile
FROM rocker/r-ver:4.2.1

ENV S6_VERSION=v2.1.0.2
ENV RSTUDIO_VERSION=daily
ENV DEFAULT_USER=rstudio
ENV PANDOC_VERSION=default
ENV PATH=/usr/lib/rstudio-server/bin:$PATH

RUN /rocker_scripts/install_rstudio.sh
RUN /rocker_scripts/install_pandoc.sh

EXPOSE 8787

CMD ["/init"]
```

![image](https://user-images.githubusercontent.com/50911393/186281510-6449fc5b-1fb4-4dc9-8eb1-cb8066fbd2a4.png)

**Comment by yufree** [comment]
Thanks for @eitsupi 's post and now the rocker/rstudio can work on arm64.

However, If I add `/rocker_scripts/install_tidyverse.sh` to the dockerfile above, the compile will fail due to two packages ( `arrow` and `duckdb` ) of dplyr database backends. I removed those two packages and add `libharfbuzz-dev` `libfribidi-dev` `libfreetype6-dev` `libpng-dev` `libtiff5-dev` `libjpeg-dev` packages, the tidyverse and verse version can work. I wonder if it's possible to add options to skip certain backends? 

Meanwhile, the `/rocker_scripts/install_quarto.sh` will also fail. 

Here is an example [Dockerfile](https://github.com/yufree/democode/blob/master/rreport/Dockerfile) with `tidyverse` supported.

**Comment by benz0li** [comment]
> However, If I add `/rocker_scripts/install_tidyverse.sh` to the dockerfile above, the compile will fail due to two packages ( `arrow` and `duckdb` ) of dplyr database backends.

---

> Will be fixed with the next minor release of R (4.2.2) as `*PICFLAGS` are now set to `-fPIC` for aarch64.

– https://github.com/duckdb/duckdb/issues/3049#issuecomment-1207162023

Cross reference: https://bugs.r-project.org/show_bug.cgi?id=18326

**Comment by benz0li** [comment]
> Meanwhile, the `/rocker_scripts/install_quarto.sh` will also fail.

---

> RStudio is now has experimental builds for ARM architectures, currently w/o bundled Quarto due to this issue. https://dailies.rstudio.com/rstudio/elsbeth-geranium/server/jammy-arm64/

– https://github.com/quarto-dev/quarto-cli/issues/190#issuecomment-1221008095

---

Cross reference: https://github.com/denoland/deno/issues/1846

**Comment by eitsupi** [comment]
@yufree
Thanks for sharing the example Dockerfile, as @benz0li pointed out, currently `install_tidyverse.sh` and `install_quarto.sh` cannot run on arm64.

1. Cannot build the `duckdb` package on arm64 platform due to R problem (will be fixed in R 4.3.0)
2. Building the `arrow` package is very hard and cannot be completed on a Raspberry Pi 4 due to machine spec as far as I have tested.
    There is also a problem that `install_tidyverse.sh` lacks `cmake` installation, so older arrow versions try to install the amd64 version of cmake even on arm64.
3. quarto cli do not have arm64 builds, so it cannot be installed.

I think we need to document which Rocker scripts are executable on arm64. (#405)

> I wonder if it's possible to add options to skip certain backends?

I agree that there is a need to skip arrow and duckdb as they are time consuming or impossible to build.
However, since it is up to the user to decide which R packages they wants to install, the option I currently consider promising is to skip installation of all R packages.
https://github.com/rocker-org/rocker-versioned2/pull/312#issuecomment-1086858951

**Comment by Dschaykib** [comment]
@eitsupi I tried [your script](https://github.com/rocker-org/rocker-versioned2/issues/144#issuecomment-1224978820) on two different macs, but in both cases, after the login, I receive an error "unable to connect to service" (the same as mentioned [here](https://github.com/rstudio/rstudio/issues/10927)). Do you have any suggestions on how to fix this or why it happend?

**Comment by Dschaykib** [comment]
I did check and I use the arm64 image.  I even included a `--platform linux/arm64` flag in the docker build, but I still get the same error.

**Comment by eitsupi** [comment]
@Dschaykib Unfortunately, there does not seem to be anything else I can advise you on the basis of the little information you provide.

**Comment by maharshi14** [comment]
Probably a stupid question but if I use a container running ubuntu, install docker and then try and use a container `rocker/rstudio` in it, would it work?

**Comment by eddelbuettel** [comment]
You can run docker inside docker but that is generally more involved.  So in short for what you write above, you would preferably _start_ with the rocker/rstudio container as it already extends (and contains) an underlying container like base ubuntu.  So aim at just one.

**Comment by eitsupi** [comment]
Hi, recent work (#578 and #584) has made `rocker/rstudio:latest-daily` ("latest" R plus "daily" build RStudio Server) multi-platform builds for amd64 and arm64.
If you are not afraid of Daily builds, give it a try.

![image](https://user-images.githubusercontent.com/50911393/209311724-5488f4c4-0c59-4fc8-b314-dd842b9f37e7.png)

**Comment by gorgitko** [comment]
Hi, excellent work with the `latest-daily` image. Can I ask when the `arm64` will be available also for versioned tags? Thanks!

**Comment by eitsupi** [comment]
@gorgitko Thanks for your feedback.
I think we can work on it once the arm64 build appears in the stable build of RStudio Server.

**Comment by martin-g** [comment]
@eitsupi Is the same valid for the `devel` tag or it could add arm64 support earlier ?

**Comment by eitsupi** [comment]
@martin-g The devel tag indicates that the version of R is devel, so the `rocker/rstudio:devel`'s RStudio version is the stable release version now.
And, `rocker/r-ver:devel` does not currently have arm64 builds.
So `rocker/rstudio:devel` cannot be built now.

**Comment by eitsupi** [comment]
Only a simple CI definition file update is required.

Of course we could do that, but we have not done so far because there have been no requests.
Since doing an arm64 build extends the build time of `rocker/r-ver` by over an hour, I was hesitant to do that with the devel image builds that we run daily.

**Comment by martin-g** [comment]
I see!
Well, count me as the first person to request it! :-)

**Comment by richpauloo** [comment]
First of all, major gratitude is in order for the maintainers here. This is an incredible project!

It took me a while to figure out that the `rocker/rstudio:latest-daily` was the only image that could run RStudio Server on the Apple M1. Other images with useful packages like `rocker/verse:latest-daily` won't run RStudio Server, so if you want to fire up RStudio Server and also use basic packages, the effort to build this image on top of `rstudio:latest-daily` falls to the user. 

As more people start to use Apple M1, it would reduce friction in getting people using RStudio Server in Docker to have compatible builds that just work. I understand that for daily builds, extending build time by > 1 hr slows things down. The question at hand is: is that effort better used by the CI bots, or users who will need to learn the limitations of rocker, and then build their own image?

**Comment by eitsupi** [comment]
@richpauloo Thanks for the feedback, I'm glad to hear it.

The main issues here are as follows:

- Since RSPM does not provide R binary packages for arm64, all R packages must be built from source. Combined with the slowdown caused by arm64 emulation, this take a very long time.
- The `duckdb` package included in `install_tidyvese.sh` cannot be installed on arm64 Linux due to a bug in R ([expected to be fixed in R 4.3](https://bugs.r-project.org/show_bug.cgi?id=18326)). 

In other words, `rocker/tidyverse:latest-daily` is currently not buildable.
arm mac can build much faster than GitHub Actions, so you should be able to build it, except for duckdb.

So if we rewrite this script and configure it to skip duckdb on arm64, we should be able to run this script at the user's hand.

https://github.com/rocker-org/rocker-versioned2/blob/23b6961dcc187b7290b35ae4434180dfb1fa7f24/scripts/install_tidyverse.sh#L51

**Comment by benz0li** [comment]
@richpauloo See https://github.com/duckdb/duckdb/issues/3049#issuecomment-1096671708 for a workaround to build `duckdb` on `aarch64` (`arm64`).

**Comment by jashapiro** [comment]
As of Docker Desktop for Mac `4.16.2`, there is now an option to run `amd64` images in Rosetta. This seems to allow older versions (and current versions) of `rocker/rstudio` to run without modification. It should also be much faster than the previous emulation. 

The option is found in the "Features in Development" section of Settings: 
<img width="571" alt="Screenshot 2023-01-24 at 9 04 30 AM" src="https://user-images.githubusercontent.com/20483/214315258-be82c0c8-53af-46de-9354-84c39187fd5a.png">

**Comment by hughcross** [comment]
@jashapiro 

Thank you for the tip. However, I just upgraded Docker Desktop to 4.16.2 and I do not see that option under "Features in development"

**Comment by echuber2** [comment]
I'm not on a Mac right now to check, but someone suggested running the updater again and restarting Docker Desktop to be sure, and check General > "Use virtualization framework" first.

**Comment by hughcross** [comment]
Thanks @echuber2 I appreciate the help. No dice, though. I updated and restarted a couple of times (with "Use virtualization framework" checked) and still do not see the Rosetta option. I also checked and I have the Apple chip version of Docker Desktop running, and Rosetta is installed. Weird.

**Comment by nikdata** [comment]
@hughcross I actually have the free plan and I can see it. Maybe a complete uninstall and reinstall may work?

<img width="1382" alt="Screenshot 2023-01-25 at 8 32 17 PM" src="https://user-images.githubusercontent.com/24734983/214745515-d15eda6e-a246-45af-9fbb-bd6c02c5b1f7.png">

**Comment by hughcross** [comment]
Thanks @nikdata I will give it a try.

**Comment by richpauloo** [comment]
> I'm not on a Mac right now to check, but someone suggested running the updater again and restarting Docker Desktop to be sure, and check General > "Use virtualization framework" first.

I would beware here, as this leads to a [well known fatal error on docker start on M1](https://github.com/docker/for-mac/issues/6487). To undo, you can always edit the system-level config file:

```
open ~/Library/Group\ Containers/group.com.docker/settings.json
```

And ensure:

```
"useVirtualizationFramework": false,
"useVirtualizationFrameworkVirtioFS": false,
"useVirtualizationFrameworkRosetta": false,
```

Then save and restart Docker Desktop.

**Comment by chendaniely** [comment]
I can confirm that the solution from @nikdata works. Tested this out with a few students in a class I'm teaching.
There's a few things you need to do to get it working:

1. Get the latest version of docker desktop
2. You **need** to update OSX to Ventura 13+ (anything older won't work)
3. Enable rosetta from command line: `softwareupdate --install-rosetta`

Then you will see the checkbox under the experimental settings.

**Comment by Martins6** [comment]
I'm using Mac with M2 chip. You guys helped a lot as well, thanks!

Just to complement the answer from @chendaniely, I had to specify the platform argument as well. The final results is the following to make it work, hope it helps someone!

```
docker run --rm \
  --platform linux/x86_64 \
  -p 8787:8787 \
  -e PASSWORD=YOURNEWPASSWORD \
  -e DISABLE_AUTH=true \
  rocker/rstudio
```

**Comment by AdamCiuris** [comment]
> I can confirm that the solution from @nikdata works. Tested this out with a few students in a class I'm teaching. There's a few things you need to do to get it working:
> 
> 1. Get the latest version of docker desktop
> 2. You **need** to update OSX to Ventura 13+ (anything older won't work)
> 3. Enable rosetta from command line: `softwareupdate --install-rosetta`
> 
> Then you will see the checkbox under the experimental settings.
> 
> Run the image:
> 
> ```
> docker run --rm \
>   -p 127.0.0.1:8787:8787 \
>   -e DISABLE_AUTH=true \
>   rocker/rstudio
> ```
> 
> Because it's doing things through Rosetta. I haven't confirmed how packages and installations + compiling are handled. 🤞

Thank you! We spent two days on this and this fixed it. Remember to update docker and tick the two settings in docker desktop that let you use rosetta!

**Comment by hughcross** [comment]
Thanks to everyone for their input and testing. I have not updated to Ventura yet (work computer, so have to go by their schedule), but once I do I will check this out.

**Comment by SameOldSamOld** [comment]
This thread has been extremely useful for me on my Apple M2. Thank you very much.

Below is the fix that finally helped me build my Dockerfile successfully!

`docker build --platform linux/x86_64 -t my_container:1.0 .`

**Comment by chendaniely** [comment]
I edited my [original comment](https://github.com/rocker-org/rocker-versioned2/issues/144#issuecomment-1412330338). It seems that for many of my students and me, we need to use `:4.1.3` right now.

Try that version if `:latest` or `:4.2.2` isn't working.

**Comment by eitsupi** [comment]
FYI, Quarto CLI 1.3 releases (now prerelease) include deb packages for arm64, so we will probably be able to build `rocker/rstudio` for arm64 in its complete form soon.

**Comment by eitsupi** [comment]
I have created a PR #651 to enable arm64 builds of `rocker/rstudio` with RStudio 2023.05.1 or later.
Ref: rstudio/rstudio#12411

**Comment by diazrenata** [comment]
Thank you, @eitsupi, for incredible work on this (going back years)! I understand if this may not be possible, but I am wondering if it would be feasible to release versions of `rocker/tidyverse` that are also arm64 compatible? It would be incredibly helpful, but I completely understand if either time or the upstream dependencies and build times are prohibitive!

**Comment by eitsupi** [comment]
`rocker/rstudio:latest` arm64 image has been pushed to DockerHub today.
![image](https://github.com/rocker-org/rocker-versioned2/assets/50911393/15cee61b-8bb3-4abb-bb0e-fa677846714f)
